Steam login with `GetAuthTicketForWebApi `

arthur740212 opened the pull request:
This will be the only way to login in with steam session tickets in the future, 
according to https://dev.epicgames.com/docs/dev-portal/identity-provider-management#steam

GetAuthTicketForWebApi is async and requires a callback.

Current state : 
First login will attempt to grab the ticket.
Will log a message when the ticket is retrieved.
Press login button again to successfully login.

Todo : 
Support custom network identity with configs

arthur740212 committed:
feat(steam) : log in with GetAuthTicketForWebApi

This will be the only way to login in with steam session tickets in the future,
according to https://dev.epicgames.com/docs/dev-portal/identity-provider-management#steam

paulhazen commented on Assets/Scripts/Steam/SteamManager.cs:
```diff
@@ -323,6 +339,7 @@ public string GetSessionTicket()
                 return sessionTicketString;
             }
 
+#if GET_STEAM_SESSION_TICKET_LEGACY
```
We should add or include documentation making clear what this scripting define does (ie why do we have this conditional here?)

paulhazen commented on Assets/Scripts/Steam/SteamManager.cs:
```diff
@@ -63,6 +63,8 @@ protected static void SteamAPIDebugTextHook(int nSeverity, System.Text.StringBui
         HAuthTicket sessionTicketHandle = HAuthTicket.Invalid;
         string sessionTicketString = null;
 
+        Callback<GetTicketForWebApiResponse_t> m_AuthTicketForWebApiResponseCallback;
```
While the rest of this file makes use in some cases of the prefix `m_`, adding _new_ things within this project we should make sure that new additions match the overall project style guidelines (at least in this case).

paulhazen reviewed (changes requested):
Requesting changes on some minor code formatting and documentation things - but I am uncertain about the nature of half-way addressing the linked issue - I'd like to understand better whether or not a complete solution can be accomplished - and why or why not in either case.

WispyMouse commented on Assets/Scripts/Steam/SteamManager.cs:
```diff
@@ -63,6 +63,8 @@ protected static void SteamAPIDebugTextHook(int nSeverity, System.Text.StringBui
         HAuthTicket sessionTicketHandle = HAuthTicket.Invalid;
         string sessionTicketString = null;
 
+        Callback<GetTicketForWebApiResponse_t> m_AuthTicketForWebApiResponseCallback;
```
Because this is a MonoBehaviour, I really want everything to have accessors. Otherwise Unity might do something funny and serialize the value of this, even though it is private. Do you think that applies here?

arthur740212 replied:
Just learned a lot about serialization, after a crash course with Mint.

SteamManager only gets created a singleton-ish instance in runtime, and there are no other references in scenes, so there should be nothing getting serialized here.
Also `Callback` here is an object from Steamworks, which Unity wouldn't be able to serialize.

WispyMouse commented on Assets/Scripts/Steam/SteamManager.cs:
```diff
@@ -148,6 +150,20 @@ protected virtual void OnEnable()
+                        {
+                            sessionTicketHandle = pCallback.m_hAuthTicket;
+                            sessionTicketString = System.BitConverter.ToString(pCallback.m_rgubTicket).Replace("-", "");
+                            Debug.LogWarning ("GetAuthTicketForWebApi succeed, try logging in again");
```
Should this use one of our conditional print functions instead of Unity's logging directly?

arthur740212 replied:
print() logs a log not a log warning.
I made this warning to stand out (warnings are yellow), so user knows the ticket is ready to try logging in again.
We decided that this two-step login will need to be changed.
So this line will probably disappear.

WispyMouse commented on Assets/Scripts/Steam/SteamManager.cs:
```diff
@@ -148,6 +150,20 @@ protected virtual void OnEnable()
+                m_AuthTicketForWebApiResponseCallback = Callback<GetTicketForWebApiResponse_t>.Create(
+                    (GetTicketForWebApiResponse_t pCallback) =>
+                    {
+                        if (pCallback.m_eResult == EResult.k_EResultOK)
```
If the result is not a success, should we log out an error message?

WispyMouse reviewed (changes requested):
I'd like to see some additional logging changes.

arthur740212 committed:
chore(optimize) : Ingest feedbacks from PR#726

Rename variable to fit naming convention
Error logs when session ticket fails to retrieve
Flip ifdef to use old session ticket grabbing method as default

arthur740212 committed:
fix(steam,login) : login with session ticket async without pressing login twice

arthur740212 committed:
fix(steam) : ifdef to select session ticket methods

andrew-hirata-playeveryware commented on Assets/Scripts/Steam/SteamManager.cs:
```diff
@@ -126,8 +129,7 @@ private void OnApplicationQuit()
 
-
-
+        TaskCompletionSource<string> tcs;
```
This should be moved to near the definition of the callback. and tcs should be renamed to something like authTicketForWebAPIResponseTaskCompletionSource or something shorter if you can think of it.

andrew-hirata-playeveryware commented on Assets/Scripts/Steam/SteamManager.cs:
```diff
@@ -312,7 +332,29 @@ public string GetSteamID()
+            // For now (2024/06/19) it still used the old way to grab session tickets
+            // Manually define GET_SESSION_TICKET_FOR_WEB_API to use the new version
+    #if ASYNC_GET_STEAM_SESSION_TICKET_PREVIEW
+            //if (sessionTicketString != null)
```
If this isn't needed, we should delete it.

arthur740212 committed:
chore(style) : rename variables, remove unused comments
